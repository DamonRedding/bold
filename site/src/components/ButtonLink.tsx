import { Button, ButtonProps } from 'bold-ui'
import GatsbyLink, { GatsbyLinkProps } from 'gatsby-link'
import React from 'react'

export interface ButtonLinkProps extends ButtonProps {
  to?: GatsbyLinkProps<any>['to']
  href?: string
}

export function ButtonLink(props: ButtonLinkProps) {
  if (props.to) {
    return <Button component={GatsbyLink} {...props} />
  } else {
    return <Button component='a' {...props} />
  }
}
