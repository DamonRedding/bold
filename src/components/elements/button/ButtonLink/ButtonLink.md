Botão que renderiza um link com a tag `<a>` ao invés de um `<button>`.
