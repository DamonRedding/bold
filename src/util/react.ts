import { MutableRefObject, Ref } from 'react'

export function composeRefs<T>(...refs: Array<Ref<T>>) {
    return (value: T) => refs.forEach(ref => setRef(ref, value))
}

export function setRef<T>(ref: Ref<T>, value: any) {
    if (typeof ref === 'function') {
        ref(value)
    } else if (ref) {
        (ref as MutableRefObject<T>).current = value
    }
}

/**
 * Returns the next sibling of `element` that matches the `predicate`.
 * Returns `null` if not found.
 */
export function getNextSibling(element: Element, predicate: (sibling: Element) => boolean) {
    let current: Element = element

    while (current.nextElementSibling) {
        current = current.nextElementSibling

        if (predicate(current)) {
            return current
        }
    }

    return null
}

/**
 * Returns the previous sibling of `element` that matches the `predicate`.
 * Returns `null` if not found.
 */
export function getPreviousSibling(element: Element, predicate: (sibling: Element) => boolean) {
    let current: Element = element

    while (current.previousElementSibling) {
        current = current.previousElementSibling

        if (predicate(current)) {
            return current
        }
    }

    return null
}
