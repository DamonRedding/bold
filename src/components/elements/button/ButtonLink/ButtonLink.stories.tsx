import { select, text, withKnobs } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/react'
import * as React from 'react'
import { MemoryRouter } from 'react-router'

import { withPropTypes } from '../../../../stories-addons/withPropTypes'
import { withTheme } from '../../../../stories-addons/withTheme'
import { Type } from '../Button/Button'

import { ButtonLink } from './ButtonLink'

const typeOptions: {[key in Type]: string} = {
    'normal': 'normal',
    'primary': 'primary',
}

storiesOf('Components', module)
    .addDecorator(withPropTypes(`
    Botão que renderiza um link com a tag <a> ao invés de um <button>.
    `))
    .addDecorator(withKnobs)
    .addDecorator(withTheme())
    .add('ButtonLink', () => (
        <MemoryRouter initialEntries={['/']} initialIndex={0}>
            <ButtonLink
                to={text('to', '/')}
                label={text('label', 'LINK')}
                type={select('type', typeOptions, 'normal')}
            />
        </MemoryRouter>
    ))
