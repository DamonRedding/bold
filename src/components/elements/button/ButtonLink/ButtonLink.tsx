import * as React from 'react'
import { Link, LinkProps } from 'react-router-dom'

import { withStyles, WithStylesProps } from '../../../../styles'
import { Type } from '../Button/Button'
import { createStyles } from '../Button/skins/default'

export interface ButtonLinkProps extends WithStylesProps, Pick<LinkProps, 'to' | 'replace'> {
    label: string
    type?: Type
}

@withStyles
export class ButtonLink extends React.PureComponent<ButtonLinkProps> {
    render() {
        const { css, theme, label, type, ...rest } = this.props
        const styles = {
            ...createStyles(theme),
            link: {
                textDecoration: 'none',
            },
        }
        const classes = css(
            styles.button,
            styles.link,
            type === 'primary' && styles.primary
        )

        return (
            <Link className={classes} {...rest}>
                {label}
            </Link>
        )
    }
}
