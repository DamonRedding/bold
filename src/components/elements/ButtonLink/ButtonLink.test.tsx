import { mount, render, shallow } from 'enzyme'
import * as React from 'react'
import { MemoryRouter } from 'react-router'

import { withTheme } from '../../../test'

import { ButtonLink } from './ButtonLink'

it('deve renderizar corretamente', () => {
    const wrapper = render(withTheme(
        <MemoryRouter initialEntries={['/']} initialIndex={0} >
            <ButtonLink to='/' label='Link to home' />
        </MemoryRouter >
    ))
    expect(wrapper).toMatchSnapshot()
})
