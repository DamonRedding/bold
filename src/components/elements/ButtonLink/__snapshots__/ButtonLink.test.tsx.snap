// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deve renderizar corretamente 1`] = `
.emotion-0 {
  background-color: #ffffff;
  border: 1px solid #b3b3b3;
  border-radius: 4px;
  color: #808080;
  cursor: pointer;
  display: inline-block;
  font-family: "Noto Sans",sans-serif;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  font-size: 0.75rem;
  font-weight: bold;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  padding: 1rem 2.5rem;
  text-transform: uppercase;
  -webkit-transition: all .2s;
  transition: all .2s;
  text-decoration: none;
}

.emotion-0 span {
  -webkit-transition: all .2s;
  transition: all .2s;
}

.emotion-0:not(:disabled):active {
  box-shadow: inset 0 2px 8px 0 rgba(0,0,0,.1);
}

.emotion-0:disabled {
  opacity: 0.5;
}

.emotion-0:focus {
  border: 1px solid #056DFF;
  outline: none;
}

<a
  class="emotion-0"
  href="/"
>
  Link to home
</a>
`;
